sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/application-changes-report-integration/_tests_/application-changes-report-integration.scheduler.e2e-spec.ts

```diff
@@ -0,0 +1,383 @@
+      await db.application.save(application);
+
+      // Queued job.
+      const mockedJob = mockBullJob<void>();
+
+      // Act
+      const processingResult = await processor.processApplicationChanges(
+        mockedJob.job,
+      );
+      // Assert
+      // Assert process result.
+      expect(processingResult).toContain("Process finalized with success.");
+      expect(processingResult).toContain("Applications reported: 1");
+      expect(
+        mockedJob.containLogMessages([
+          "Found 1 application changes.",
+          "Reported date has been successfully updated for reported application assessments.",
+        ]),
+      ).toBe(true);
+      const uploadedFile = getUploadedFile(sftpClientMock);
+      const expectedFirstRecord = getExpectedApplicationChangesCSVRecord(
+        application,
+        "PT",
+        "Reassessment",
+      );
+      // Assert file output.
+      const [header, firstRecord, numberOfRecords] = uploadedFile.fileLines;
+      expect(header).toBe(
+        Object.values(ApplicationChangesReportHeaders).join(","),
+      );
+      expect(firstRecord).toBe(expectedFirstRecord);
+      expect(numberOfRecords).toBe("Number of records: 1");
+      // Assert the report date.
+      const reportedAssessment = await db.studentAssessment.findOne({
+        select: { id: true, reportedDate: true },
+        where: {
+          id: application.currentAssessment.id,
+        },
+      });
+      expect(reportedAssessment.reportedDate).toBeInstanceOf(Date);
+    });
+
+    it("Should generate blank application changes report when there is no application changes to be reported.", async () => {
+      // Arrange
+      // Preparing application which has no application changes to be reported as the reported date is set.
+      await createApplicationChangesData(
+        sharedStudent,
+        1,
+        -1,
+        AssessmentTriggerType.OfferingChange,
+        true,
+      );
+
+      // Queued job.
+      const mockedJob = mockBullJob<void>();
+
+      // Act
+      const processingResult = await processor.processApplicationChanges(
+        mockedJob.job,
+      );
+      // Assert
+      // Assert process result.
+      expect(processingResult).toContain("Process finalized with success.");
+      expect(processingResult).toContain("Applications reported: 0");
+      expect(
+        mockedJob.containLogMessages([
+          "Found 0 application changes.",
+          "Report date update not required as no application changes are reported.",
+        ]),
+      ).toBe(true);
+      const uploadedFile = getUploadedFile(sftpClientMock);
+      // Assert file output.
+      const [header, firstRecord, numberOfRecords] = uploadedFile.fileLines;
+      expect(header).toBe(
+        Object.values(ApplicationChangesReportHeaders).join(","),
+      );
+      // A blank record file is expected to be generated when there is no application changes reported.
+      expect(firstRecord).toBe("");
```

Comment:
For clarification, is the empty line a decision or is it just the CSV lib output when there are no records?

Reply:
That is a CSV lib output when there are no records. (i.e Header line + EOL + empty string)